plugins/module_utils/vmware_nsxt.py

```diff
@@ -44,62 +45,76 @@ def request(url, data=None, headers=None, method='GET', use_proxy=True,
+        validate_certs, url_username, url_password, http_agent,
+        force_basic_auth, ignore_errors, client_cert)
+
+def get_all_results(
+        url, data, headers, method, use_proxy, force, last_mod_time, timeout,
+        validate_certs, url_username, url_password, http_agent,
+        force_basic_auth, ignore_errors, client_cert):
     NULL_CURSOR_PREFIX = '0000'
+    rc, resp = _request(
+        url, data, headers, method, use_proxy, force, last_mod_time, timeout,
+        validate_certs, url_username, url_password, http_agent,
+        force_basic_auth, ignore_errors, client_cert)
+    if rc != 200:
+        return rc, None
+    cursor = resp.get('cursor', NULL_CURSOR_PREFIX)
+    op = '&' if urlparse.urlparse(url).query else '?'
+    url += op + 'cursor='
+    while cursor and not cursor.startswith(NULL_CURSOR_PREFIX):
+        rc, page = _request(
+            url + cursor, data, headers, method, use_proxy, force,
+            last_mod_time, timeout, validate_certs, url_username,
+            url_password, http_agent, force_basic_auth, ignore_errors,
+            client_cert)
+        if rc != 200:
+            return rc, None
+        resp['results'].extend(page.get('results', []))
+        cursor = page.get('cursor', NULL_CURSOR_PREFIX)
+    return rc, resp
+
+def _request(url, data, headers, method, use_proxy,
+             force, last_mod_time, timeout, validate_certs,
+             url_username, url_password, http_agent, force_basic_auth,
+             ignore_errors, client_cert):
+    ca_path = get_certificate_file_path('NSX_MANAGER_CA_PATH')
     resp_data = None
-    page = None
     try:
-        r = open_url(url=url, data=data, headers=headers, method=method, use_proxy=use_proxy,
-                     force=force, last_mod_time=last_mod_time, timeout=timeout, validate_certs=validate_certs,
-                     url_username=url_username, url_password=url_password, http_agent=http_agent,
-                     client_cert=client_cert, force_basic_auth=force_basic_auth, ca_path=ca_path)
+        r = open_url(
+            url=url, data=data, headers=headers, method=method,
+            use_proxy=use_proxy, force=force, last_mod_time=last_mod_time,
+            timeout=timeout, validate_certs=validate_certs,
+            url_username=url_username, url_password=url_password,
+            http_agent=http_agent, client_cert=client_cert,
+            force_basic_auth=force_basic_auth, ca_path=ca_path)
     except HTTPError as err:
         r = err
 
     try:
-        raw_data = r.read()
+        raw_data = r.read().decode('utf-8')
         if raw_data:
             if is_json(raw_data):
                 resp_data = json.loads(raw_data)
             else:
                 resp_data = raw_data
     except Exception as e:
-        if ignore_errors:
-            pass
-        else:
+        if not ignore_errors:
             raise Exception(e)
 
     resp_code = r.getcode()
 
     if resp_code >= 400 and not ignore_errors:
```

Comment:
As we are not handling 3xx response codes should we treating them as errors?

Reply:
Refer https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py#L982

I think that we won't hit 3xx errors because we're using follow_redirects='urllib2' (i.e. default value of this param).

What do you reckon?

Reply:
I think you're right. I did not consider we were using open_url from ansible's module_utils